project/Bootstrap.scala

```diff
@@ -1,55 +0,0 @@
-package sbt
-
-import sbt.internal.classpath.ClassLoaderCache
-import sbt.internal.inc.ScalaInstance
-
-// This class needs to be in package sbt to access the ClassLoaderCache
-object Bootstrap {
-  /** A ScalaInstance (without scaladoc). */
-  def makeScalaInstance(
-    state: State,
-    version: String,
-    libraryJars: Array[File],
-    compilerJars: Array[File],
-    topLoader: ClassLoader
-  ): ScalaInstance = {
-    // `extendedClassLoaderCache` is package private in package sbt
-    val cache = state.extendedClassLoaderCache
-
-    val libraryLoader = cache(libraryJars.toList, topLoader)
-    val compilerLoader = cache(compilerJars.toList, libraryLoader)
-
-    new ScalaInstance(
-      version = version,
-      loader = compilerLoader,
-      loaderCompilerOnly = compilerLoader,
-      loaderLibraryOnly = libraryLoader,
-      libraryJars = libraryJars,
-      compilerJars = compilerJars,
-      allJars = libraryJars ++ compilerJars,
-      explicitActual = Some(version)
-    )
-  }
-
-  /** A ScalaInstance identical to `base` but with additional jars for scaladoc. */
-  def makeDocScalaInstance(
-    state: State,
-    base: ScalaInstance,
-    docJars: Array[File]
-  ): ScalaInstance = {
-    val cache = state.extendedClassLoaderCache
-
-    val fullLoader = cache(docJars.toList, base.loaderCompilerOnly)
```

Comment:
This version reused the classloader from base for compiler and library jars, whereas the new one will recreate it it seems, isn't that likely to be slower and require more memory?

Reply:
Class loaders are cached internally by sbt, based on the jars file. Here we use compiler jars from the base Scala Instance, so the compiler class loader should be the same instance.

Reply:
Actually there is something wrong the list of compiler jars, so I will fix it and add an assert to ensure that the class loaders are the same.

Reply:
The problem is https://github.com/sbt/sbt/blob/fabeed222799efcc46be15bccf2c31e384f7e369/main/src/main/scala/sbt/Defaults.scala#L1186-L1187

It's like a snake biting it's own tail...